clean_code_partner_review

khanh-devos opened the pull request:
#### This project is list all tasks (todos) that need be done  within a week. It use mainly JS and webpack to build. 
It has a class Task and data.js:
- class Tasks includes methods: addNew, deleteArrTask, deleteSingleTask, enableEditTask
- data.js includes methods dealing with saving or deleting tasks into localStorage.

khanh-devos committed:
fix webhint err

khanh-devos committed:
fix webhint err

manzitresor commented on src/tasks.js:
```diff
@@ -0,0 +1,195 @@
+    this.refreshTasks();
+  }
+
+  enableEdit = (e) => {
```
- [ ] [SUGGESTED] Kindly Replace the arrow function syntax **enableEdit = (e) => { }** with the regular function syntax **enableEdit(e) { }**. This change will ensure better compatibility with all environments and avoid potential

manzitresor replied:
- [ ] [SUGGESTED] Kindly Replace the arrow function syntax **enableEdit = (e) => { }** with the regular function syntax **enableEdit(e) { }.** This change will ensure better compatibility with all environments and avoid potential issues.

manzitresor reviewed (commented):
Hi @khanh-devos 
![image](https://github.com/khanh-devos/TODO_LIST_clean_code_partner_review/assets/83731517/001b9392-506f-495e-ad9a-f9ce2439453b)


Great job so far! 👏
You've done well  By well implementing project required , however, there are some issues that you still need to work on to make your code look more professional

manzitresor commented on src/data.js:
```diff
@@ -0,0 +1,55 @@
+  tasks[idx1 - 1] = task2;
+  tasks[idx2 - 1] = task1;
+  setData(tasks);
+};
```
- [x] [SUGGESTED]Consider making the **exhangeData** function more descriptive and fixing the typo. Update the name to **exchangeData**. Additionally, modify the code to use array destructuring for better readability:

manzitresor commented on src/tasks.js:
```diff
@@ -0,0 +1,195 @@
+import {
+  deleteArrOfIndex, editTask, exhangeData, getData, setData,
+} from './data.js';
```
- [ ] [SUGGESTED]Instead of importing individual functions from **'./data.js'**, consider using a wildcard import (*) to import all functions from the module. This ensures that any future additions or changes to the module will be included automatically.

manzitresor commented on src/tasks.js:
```diff
@@ -0,0 +1,195 @@
+    this.draggedLiId = null;
+  }
+
+  buildForm1 = () => {
```
- [ ] [SUGGESTED]Replace the arrow function syntax **buildForm1 = () => { }** with the regular function syntax **buildForm1() { }.** This change will ensure better compatibility with all environments and avoid potential issues.

khanh-devos replied:
Hello Manzi, 
Thank you for your suggestion, can you give more detailed information for me to get the difference between using regular or arrow _functions_ inside a class ?

manzitresor replied:
Hi@khanh-devos 
Thank you for asking Here's a more detailed explanation of the difference between using regular and arrow functions inside a class:

- [ ] When using regular function syntax, you define the method as buildForm1() { }. This means that the method is a regular function that will be hoisted and has its own **this** context. It allows you to access the class instance properties and call other methods on the class instance. Regular functions are more flexible and can be useful.
- [ ] On the other hand, arrow function syntax is buildForm1 = () => { }. Arrow functions do not have their own this context and instead inherit this from the surrounding lexical scope .When used as a method in a class, the arrow function will capture the this value of the class instance.

In your case, I suggested using regular function syntax buildForm1() { } instead of arrow function syntax 
buildForm1 = () => { } for better compatibility with all environments and to avoid potential issues.

khanh-devos replied:
Hello @Manzi,
Thank you for your quick response.
Your answere is exactly what I wanted from the beginning:

- [ ] Arrow function does not have its own "this". This will allow me to use this.otherFunction() inside its block very easily, I do not need to use method bind() inside the class constructor.
- [ ] And also the instance todos = new Task(), that can also call any **arrow or regular functions** declared inside the class, you can see it at index.js

manzitresor replied:
Thank you for your feedback, @khanh-devos. I'm glad to hear that my previous response was helpful to you. 🤝 

 

- [ ] As you mentioned, since you're not using bind() inside the constructor and the arrow functions are serving your needs effectively, you can continue to use arrow functions in your code.

khanh-devos replied:
Hello @Manzi, 
Thank you for your best-practise suggestions, 
- I've checked and tried many experiments so far. 
- Although it is clearly convenient to use "this" directly inside, I found out the reason that we shouldn't use arrow functions inside a class because they will be created every time an instance is created. 
- Details [here](https://www.charpeni.com/blog/arrow-functions-in-class-properties-might-not-be-as-great-as-we-think) 

#### Thank you for your suggestions.
![thanks](https://github.com/khanh-devos/TODO_LIST_clean_code_partner_review/assets/68627859/bdf4f46b-2fd1-4eba-8812-14124da77a05)

manzitresor commented on src/tasks.js:
```diff
@@ -0,0 +1,195 @@
+    document.querySelector('main').appendChild(form);
+  }
+
+  buildForm2 = () => {
```
- [ ] [SUGGESTED]Replace the arrow function syntax **buildForm2 = () => { }** with the regular function syntax **buildForm2() { }.** This change will ensure better compatibility with all environments and avoid potential issues.

manzitresor commented on src/tasks.js:
```diff
@@ -0,0 +1,195 @@
+    document.querySelector('main').appendChild(form);
+  }
+
+  liTemplate = (task) => {
```
- [ ] [SUGGESTED]Replace the arrow function syntax **liTemplate = () => { }** with the regular function syntax **liTemplate() { }.** This change will ensure better compatibility with all environments and avoid potential issues.

manzitresor commented on src/tasks.js:
```diff
@@ -0,0 +1,195 @@
+    return li;
+  }
+
+  defineLi = (e) => {
```
- [ ] [SUGGESTED] Consider Replacing the arrow function syntax **defineLi = (e) => { }** with the regular function syntax **defineLi(e) { }.** This change will ensure better compatibility with all environments and avoid potential issues.

manzitresor commented on src/tasks.js:
```diff
@@ -0,0 +1,195 @@
+    return li;
+  }
+
+  drag = (e) => {
```
- [ ] [SUGGESTED] Kindly Replace the arrow function syntax 
**drag = (e) => { }** with the regular function syntax **drag(e) { }.** This change will ensure better compatibility with all environments and avoid potential issues.

manzitresor commented on src/tasks.js:
```diff
@@ -0,0 +1,195 @@
+    this.draggedLiId = e.target.id;
+  }
+
+  dragover = (e) => {
```
- [ ] [SUGGESTED] Kindly Replace the arrow function syntax **dragover = (e) => { }** with the regular function syntax **dragover(e) { }.** This change will ensure better compatibility with all environments and avoid potential issues.

manzitresor commented on src/tasks.js:
```diff
@@ -0,0 +1,195 @@
+    e.preventDefault();
+  }
+
+  drop = (e) => {
```
- [ ] [SUGGESTED] Consider Replacing the arrow function syntax 
**drop = (e) => { }** with the regular function syntax **drop(e) { }.** This change will ensure better compatibility with all environments and avoid potential issues.

manzitresor commented on src/tasks.js:
```diff
@@ -0,0 +1,195 @@
+    }
+  }
+
+  refreshTasks = () => {
```
- [ ] [SUGGESTED]consider Replacing the arrow function syntax **refershTasks = () => { }** with the regular function syntax **refreshTasks() { }.** This change will ensure better compatibility with all environments and avoid potential issues.

manzitresor commented on src/tasks.js:
```diff
@@ -0,0 +1,195 @@
+    this.showTasks();
+  }
+
+  showTasks = () => {
```
- [ ] [SUGGESTED] Kindly Replace the arrow function syntax **showTasks= () => { }** with the regular function syntax **showTasks() { }.** This change will ensure better compatibility with all environments and avoid potential issues.

manzitresor commented on src/tasks.js:
```diff
@@ -0,0 +1,195 @@
+    document.querySelector('#form2').insertAdjacentElement('afterbegin', ul);
+  }
+
+  addNew = (e) => {
```
- [ ] [SUGGESTED] Kindly Replace the arrow function syntax **addNew = (e) => { }** with the regular function syntax **addNew(e) { }.** This change will ensure better compatibility with all environments and avoid potential issues.

manzitresor commented on src/tasks.js:
```diff
@@ -0,0 +1,195 @@
+    }
+  }
+
+  deleteTasks = (e) => {
```
- [ ] [SUGGESTED] Kindly Replace the arrow function syntax **deleteTasks = (e) => { }** with the regular function syntax **deleteTasks(e) { }.** This change will ensure better compatibility with all environments and avoid potential issues.

manzitresor commented on src/tasks.js:
```diff
@@ -0,0 +1,195 @@
+    this.refreshTasks();
+  }
+
+  deleteSingleTask = (e) => {
```
- [ ] [SUGGESTED] Kindly Replace the arrow function syntax **deleteSingleTask = (e) => { }** with the regular function syntax **deleteSingleTasks(e) { }.** This change will ensure better compatibility with all environments and avoid potential issues.

manzitresor reviewed (changes requested):
Hi @khanh-devos 
Great job so far! 👏

You've done well in implementing the project requirements. However, there are a few areas where you can make improvements to make your code look more professional.


![unnamed](https://github.com/khanh-devos/TODO_LIST_clean_code_partner_review/assets/83731517/5f76e002-a581-4f1c-9226-bc8333425977)

### Highlights 🎯

✅ Correct use of Github workflow
✅ No linter error
### SUGGESTED CHANGES
Check the comments under the review.

khanh-devos committed:
Follow ES6 best-practises.
- implementing destructuring at module "exchangeData"
- replace all arrow functions with regular functions in the class Task
- import * instead of many modules from data.js

khanh-devos committed:
fix linter errors

khanh-devos committed:
build webpack and modify Live Demo at Readme file

manzitresor reviewed (approved):
### Status: Approved ✔️
Hi @khanh-devos 
Your project is complete! There is nothing else to say other than... it's time to merge it :shipit:
Congratulations! 🎉
![unnamed (2)](https://github.com/khanh-devos/TODO_LIST_clean_code_partner_review/assets/83731517/7274288a-0e2e-4855-b2e6-cc7a1512ea29)


### Top highlight 🎯

- [ ] Required changes implemeted✅
